openmmtools/forcefactories.py

```diff
@@ -110,22 +121,41 @@ def restrain_atoms(thermodynamic_state, sampler_state, topology,
     K = thermodynamic_state.kT / sigma**2  # Spring constant.
     system = thermodynamic_state.system  # This is a copy.
 
-    # Make sure the topology is an MDTraj topology.
-    if isinstance(topology, mdtraj.Topology):
-        mdtraj_topology = topology
-    else:
-        mdtraj_topology = mdtraj.Topology.from_openmm(topology)
-
-    # Translate the system to the origin to avoid
-    # MonteCarloBarostat rejections (see openmm#1854).
-    protein_atoms = mdtraj_topology.select('protein')
-    distance_unit = sampler_state.positions.unit
-    centroid = np.mean(sampler_state.positions[protein_atoms,:] / distance_unit, 0) * distance_unit
-    sampler_state.positions -= centroid
+    # Check that there are atoms to restrain.
+    if len(restrained_atoms) == 0:
+        raise ValueError('No atoms to restrain.')
+
+    # We need to translate the restrained molecule to the origin
+    # to avoid MonteCarloBarostat rejections (see openmm#1854).
+    if thermodynamic_state.pressure is not None:
+        # First, determine all the molecule atoms. Reference platform is the cheapest to allocate?
+        reference_platform = openmm.Platform.getPlatformByName('Reference')
+        integrator = openmm.VerletIntegrator(1.0*unit.femtosecond)
+        context = openmm.Context(system, integrator, reference_platform)
+        molecules_atoms = context.getMolecules()
+        del context, integrator
+
+        # Make sure the atoms to restrain belong only to a single molecule.
+        molecules_atoms = [set(molecule_atoms) for molecule_atoms in molecules_atoms]
+        restrained_atoms_set = set(restrained_atoms)
+        restrained_molecule_atoms = None
+        for molecule_atoms in molecules_atoms:
+            if restrained_atoms_set.issubset(molecule_atoms):
+                # Convert set to list to use it as numpy array indices.
+                restrained_molecule_atoms = list(molecule_atoms)
+                break
+        if restrained_molecule_atoms is None:
```

Comment:
Question: Doesn't this mean that ALL restraints in YANK will not work in explicit solvent? Since the ligand is restrained relative to the receptor, then its two separate molecules and so will always fail this criteria?